frame.go

```diff
@@ -124,28 +124,41 @@ func NewFrame(payloadCapacity int) *Frame {
 	return f
 }
 
-// ReadIn reads the frame from the given io.Reader
-func (f *Frame) ReadIn(r io.Reader) error {
-	var rbuf typed.ReadBuffer
-	rbuf.Wrap(f.headerBuffer)
-
-	if _, err := rbuf.FillFrom(r, FrameHeaderSize); err != nil {
+// ReadBody takes in a previously read frame header, and only reads in the body
+// based on the size specified in the header. This allows callers to defer
+// the frame allocation till the body needs to be read.
+func (f *Frame) ReadBody(header []byte, r io.Reader) error {
+	// Copy the header into the underlying buffer so we have an assembled frame
+	// that can be directly forwarded.
+	copy(f.buffer, header)
+
+	// Parse the header into our typed struct.
+	if err := f.Header.read(typed.NewReadBuffer(header)); err != nil {
 		return err
 	}
 
-	if err := f.Header.read(&rbuf); err != nil {
-		return err
-	}
 	switch payloadSize := f.Header.PayloadSize(); {
 	case payloadSize > MaxFramePayloadSize:
 		return fmt.Errorf("invalid frame size %v", f.Header.size)
 	case payloadSize > 0:
-		if _, err := io.ReadFull(r, f.SizedPayload()); err != nil {
-			return err
-		}
+		_, err := io.ReadFull(r, f.SizedPayload())
+		return err
+	default:
+		// No payload to read
+		return nil
 	}
+}
 
-	return nil
+// ReadIn reads the frame from the given io.Reader.
+// This method is deprecated: callers should prefer to read the frame header
+// separately, and using ReadBody.
```

Comment:
Any reason not to use the "Deprecated: .." form which makes linters fail?

Reply:
👍